File: dotnet-desktop-guide/net/winforms/controls-design/snippets/how-to-create-simple-custom-control/csharp/FirstControl.cs
using System.ComponentModel;

namespace UserControlProject;

// <firstcontrol>
public partial class FirstControl : Control
{
    // <field>
    private HorizontalAlignment _textAlignment = HorizontalAlignment.Left;
    // </field>

    // <attributes>
    [System.ComponentModel.Category("Alignment"),
    System.ComponentModel.Description("Specifies the alignment of text."),
    System.ComponentModel.DefaultValue(HorizontalAlignment.Left)]
    public HorizontalAlignment TextAlignment
    // </attributes>
    {
        get => _textAlignment;
        set
        {
            _textAlignment = value;
            Invalidate();
        }
    }

    public FirstControl()
    {
        InitializeComponent();
    }

    // <ontextchanged>
    protected override void OnTextChanged(EventArgs e)
    {
        base.OnTextChanged(e);
        Invalidate();
    }
    // </ontextchanged>

    // <onpaint>
    protected override void OnPaint(PaintEventArgs pe)
    {
        // <stringformat>
        StringFormat style = new();
        // </stringformat>

        // <alignment>
        style.Alignment = TextAlignment switch
        {
            // Map the HorizontalAlignment enum to the StringAlignment enum
            HorizontalAlignment.Left => StringAlignment.Near,
            HorizontalAlignment.Right => StringAlignment.Far,
            HorizontalAlignment.Center => StringAlignment.Center,
            
            // Default to Near alignment
            _ => StringAlignment.Near
        };
        // </alignment>

        // <drawstring>
        // Create the brush and automatically dispose it.
        using SolidBrush foreBrush = new(ForeColor);

        // Call the DrawString method to write text.
        // Text, Font, and ClientRectangle are inherited properties.
        pe.Graphics.DrawString(Text, Font, foreBrush, ClientRectangle, style);
        // </drawstring>

        // <base_onpaint>
        base.OnPaint(pe);
        // </base_onpaint>
    }
    // </onpaint>
}
// </firstcontrol>

File: dotnet-desktop-guide/winforms/controls-design/snippets/how-to-create-simple-custom-control/vb/FirstControl.vb
'<firstcontrol>
Public Class FirstControl

    '<field>
    Private _textAlignment As HorizontalAlignment = HorizontalAlignment.Left
    '</field>

    '<attributes>
    <System.ComponentModel.Category("Alignment"),
    System.ComponentModel.Description("Specifies the alignment of text."),
    System.ComponentModel.DefaultValue(HorizontalAlignment.Left)>
    Public Property TextAlignment As HorizontalAlignment
    '</attributes>
        Get
            Return _textAlignment
        End Get

        Set(value As HorizontalAlignment)
            _textAlignment = value
            Invalidate()
        End Set
    End Property

    '<ontextchanged>
    Protected Overrides Sub OnTextChanged(e As EventArgs)
        MyBase.OnTextChanged(e)
        Invalidate()
    End Sub
    '</ontextchanged>

    '<onpaint>
    Protected Overrides Sub OnPaint(ByVal e As System.Windows.Forms.PaintEventArgs)
        '<stringformat>
        Dim style As New StringFormat
        '</stringformat>

        '<alignment>
        'Map the HorizontalAlignment enum to the StringAlignment enum
        Select Case TextAlignment
            Case HorizontalAlignment.Left
                style.Alignment = StringAlignment.Near
            Case HorizontalAlignment.Right
                style.Alignment = StringAlignment.Far
            Case HorizontalAlignment.Center
                style.Alignment = StringAlignment.Center
        End Select
        '</alignment>

        '<drawstring>
        'Create the brush and automatically dispose it.
        Using foreBrush As New SolidBrush(ForeColor)
            'Call the DrawString method to write text.
            'Text, Font, and ClientRectangle are inherited properties.
            e.Graphics.DrawString(Text, Font, foreBrush, ClientRectangle, style)
        End Using
        '</drawstring>
        '<base_onpaint>
        MyBase.OnPaint(e)
        '</base_onpaint>
    End Sub
    '</onpaint>

End Class
'</firstcontrol>
